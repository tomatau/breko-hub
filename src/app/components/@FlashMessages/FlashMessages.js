import { connect } from 'react-redux'
import { noop } from 'lodash'
import PropTypes from 'prop-types'
import { removeMessage } from 'app/actions/flash.actions'
import { Bem } from 'app/utils'
import * as flashSelectors from 'app/selectors/flash.selectors'
import Msg from './Msg'

@connect(state => ({
  messages: flashSelectors.getMessages(state),
}), { removeMessage })
export default class FlashMessages extends React.Component {
  static propTypes = {
    messages: PropTypes.array,
    removeMessage: PropTypes.func,
  };

  static defaultProps = {
    messages: [],
    removeMessage: noop,
  };

  static bem = new Bem('FlashMessages');

  clickMessage(msg) {
    this.props.removeMessage(msg.id)
  }

  render() {
    const { messages } = this.props
    return (
      <div {...FlashMessages.bem()}>
        {messages.map(msg =>
          <Msg key={msg.id}
            msg={msg}
            onClick={() => this.clickMessage(msg)}
            {...FlashMessages.bem('Msg')}
          />
        )}
      </div>
    )
  }
}
