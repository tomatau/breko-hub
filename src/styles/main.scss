// placeholder file
@import '~vars';
@import '~tao-scss/index';

$body-colour: #444;

body {
  color: $body-colour;
  font-family: Arial;
}
