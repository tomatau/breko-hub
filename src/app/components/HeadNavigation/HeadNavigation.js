import React from 'react'
import cx from 'classnames'
import { NavLink, withRouter } from 'react-router-dom'
import { nav as navCopy } from 'app/copy'
import { cleanProps } from 'app/utils'
import styles from './HeadNavigation.module.scss'

export default withRouter(class HeadNavigation extends React.Component {
  render() {
    const { className, ...props } = this.props
    return (
      <nav
        className={cx('HeadNavigation', styles.nav, className)}
        {...cleanProps(props)}
      >
        <NavLink
          exact
          to='/'
          activeClassName={styles.active}
        >
          {navCopy.home}
        </NavLink>
        <NavLink to='/bar' activeClassName={styles.active}>
          {navCopy.bar}
        </NavLink>
        <NavLink to='/private' activeClassName={styles.active}>
          {navCopy.private}
        </NavLink>
      </nav>
    )
  }
})
