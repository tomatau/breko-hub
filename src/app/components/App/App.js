import React from 'react'
import {Â Helmet as DocumentMeta } from 'react-helmet-async'
import { connect } from 'react-redux'
import { Switch, Route, withRouter } from 'react-router-dom'
import { hot } from 'react-hot-loader/root'
import { get } from 'app/utils'
import { addMessage } from 'app/modules/flash/flash.actions'
import { app as appCopy } from 'app/copy'
import { routesList } from 'app/routes'
import HeadNavigation from 'app/components/HeadNavigation/HeadNavigation'
import FlashMessages from 'app/components/@FlashMessages/FlashMessages'
import style from './App.module.scss'

const getLocationFlashState = get('state.flash', {})

@connect(null, { addMessage })
@withRouter
class App extends React.Component {
  componentDidMount() {
    this.addFlashFromLocation()
  }

  componentDidUpdate() {
    this.addFlashFromLocation()
  }

  addFlashFromLocation = () => {
    const { location, addMessage } = this.props
    const flashState = getLocationFlashState(location)
    if (flashState.message) {
      addMessage(flashState.message, flashState.type)
    }
  }

  render() {
    return (
      <>
        <DocumentMeta
          defaultTitle={`${appCopy.title}`}
          titleTemplate={`%s | ${appCopy.title}`}
        >
          <html lang='en' />
          <meta charSet='utf-8' />
          <meta name='description' content={appCopy.meta.description} />
          <meta name='keywords' content={appCopy.meta.keywords} />
          <meta
            name='viewport'
            content='width=device-width,initial-scale=1.0'
          />
        </DocumentMeta>
        <div className={style.app}>
          <HeadNavigation />
          <header>
            <h1>{appCopy.title}</h1>
          </header>
          <main
            aria-live='polite'
            className={style.content}
          >
            <FlashMessages />
            <Switch>
              {routesList.map(route => (
                <Route
                  key={route.name}
                  exact={route.exact}
                  path={route.path}
                  {...route.render
                    ? { render: route.render }
                    : { component: route.component }
                  }
                />
              ))}
            </Switch>
          </main>
        </div>
      </>
    )
  }
}

export default hot(App)
