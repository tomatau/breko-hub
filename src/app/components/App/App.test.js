import DocumentMeta from 'react-helmet'
import HeadNavigation from 'app/components/HeadNavigation/HeadNavigation'
import FlashMessages from 'app/components/@FlashMessages/FlashMessages'
import App from './App'
import { shallow } from 'enzyme'
import styles from './App.module.scss'
import avatarPath from 'assets/avatar.jpeg'

describe('App Component', function() {
  beforeEach(()=> {
    this.tree = shallow(<App />)
  })

  it('renders a div tag with className at rootNode', ()=> {
    expect(this.tree.hasClass(styles.app)).to.eql(true)
  })

  it('renders a Helmet document meta as firtChild', ()=> {
    const firstChild = this.tree.childAt(0)
    expect(firstChild).to.have.type(DocumentMeta)
  })

  it('renders HeadNavigation as second child', ()=> {
    const secondChild = this.tree.childAt(1)
    expect(secondChild).to.have.type(HeadNavigation)
  })

  it('renders FlashMessages as third child', ()=> {
    const thirdChild = this.tree.childAt(2)
    expect(thirdChild).to.have.type(FlashMessages)
  })

  it('renders an img as fourth child', ()=> {
    const fourthChild = this.tree.childAt(3)
    expect(fourthChild).to.have.type('img')
  })

  it('renders a h1 as fifth child', ()=> {
    const fifthChild = this.tree.childAt(4)
    expect(fifthChild).to.have.type('h1')
  })

  it('renders a main.content as sixth child', ()=> {
    const sixthChild = this.tree.childAt(5)
    expect(sixthChild).to.have.type('main')
    expect(sixthChild).to.have.className(styles.content)
  })

  describe('DocumentMeta', () => {
    it('sets the meta options', () => {
      expect(this.tree.find(DocumentMeta).node).to.eql(
        <DocumentMeta
          defaultTitle='Breko Hub'
          titleTemplate='%s | Breko Hub'>
          <html lang='en' />
          <meta charSet='utf-8' />
          <meta name='viewport' content='width=device-width,initial-scale=1.0' />
          <meta name='description' content='Breko Hub, a minimal boilerplate for building universal react applications' />
          <meta name='keywords' content='react,redux,react-router,koa,universal,babel,es7,hmr,webpack' />
        </DocumentMeta>
      )
    })
  })

  describe('Img', () => {
    it('has avatar as src with alt=me and width=70', () => {
      const img = this.tree.find('img')
      expect(img.props()).to.shallowDeepEqual({
        src: avatarPath,
        alt: 'me',
        width: '70',
      })
    })
  })

  describe('Title', () => {
    it('should have "Breko Hub" as text child', () => {
      const title = this.tree.find('h1')
      expect(title.text()).to.contain('Breko Hub')
    })
  })

  describe('Main', () => {
    it('renders the children in a main', ()=> {
      const children = <p><span>test</span><span>child</span></p>
      const content = shallow(<App>{children}</App>).find('main')

      expect(content.containsMatchingElement(children)).to.eql(true)
    })
  })
})
