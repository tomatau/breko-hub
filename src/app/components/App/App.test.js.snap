exports[`App Component Main renders routes inside Main 1`] = `
<main
  className="src-app-components-App-App-module-content">
  <Switch>
    <CodeSplitRoute
      exact={true}
      load={[Function load]}
      path="/" />
    <CodeSplitRoute
      load={[Function load]}
      path="/bar" />
    <CodeSplitRoute
      load={[Function load]}
      path="/oops" />
    <CodeSplitRoute
      load={[Function load]}
      path="/private" />
    <CodeSplitRoute
      load={[Function load]} />
  </Switch>
</main>
`;
